### PR TITLE
Fixes/options label class and attributes

### PR DESCRIPTION
**Pull request by lilHermit**
Allow FormHelper::input to pass options to ControlLabels (labels around Radio and MultiCheckbox)
    Added `optionsLabel`
Added tests
Fixes any tests broken by this

Only slight functionality change is if class was set via the direct methods (FormHelper::multiCheckbox or FormHelper::radio) and the item is checked then 'selected' gets added. Previously it was skipped

### CODE REVIEW

File: src/View/StringTemplate.php [hunk 1]
```diff
@@ -323,4 +324,65 @@ protected function _formatAttribute($key, $value, $escape = true)
...
+     * @return array|string
+     */
+    public function addClass($input, $newClass, $options = [])
+    {
+        // NOOP
+        if (empty($newClass)) {
+            return $input;
+        }
+
+        $options = $options + [
```
Comment (dereuromark) [inline src/View/StringTemplate.php]: Are you sure thats the direction you want? Having $options being higher prioritized than the second array?

File: src/View/StringTemplate.php [hunk 2]
```diff
@@ -323,4 +324,65 @@ protected function _formatAttribute($key, $value, $escape = true)
...
+        $useIndex = $options['useIndex'];
+        if (is_string($useIndex)) {
+            $class = Hash::get($input, $useIndex, []);
+        } else {
+            $class = $input;
+        }
+
+        // Convert and sanitise the inputs
+        if (!is_array($class)) {
+            if (is_string($class) && !empty($class)) {
```
Comment (dereuromark) [inline src/View/StringTemplate.php]: Do we need this? The only values allowed are array or string.

File: src/View/StringTemplate.php [hunk 3]
```diff
@@ -323,4 +324,65 @@ protected function _formatAttribute($key, $value, $escape = true)
...
+        }
+
+        if (is_string($newClass)) {
+            $newClass = explode(' ', $newClass);
+        }
+
+        $class = array_unique(array_merge($class, $newClass));
+
+        if ($options['returnArray'] === false) {
+            $class = implode(" ", $class);
```
Comment (dereuromark) [inline src/View/StringTemplate.php]: You used `'` everywhere else, would be good to keep it consistent as `' '` here, too.

File: tests/TestCase/View/StringTemplateTest.php [hunk 4]
```diff
@@ -300,4 +300,97 @@ public function testPushPopTemplates()
         $this->assertNull($this->template->pop());
         $this->assertNull($this->template->pop());
     }
+
+    public function testAddClassMethod()
+    {
+        // Test new class as null, string, array, false etc
```
Comment (dereuromark) [inline tests/TestCase/View/StringTemplateTest.php]: If you make separate test methods per "scenario", you will not need the inline comments.

testAddClassMethodNewClass
testAddClassMethodCurrentClass

etc for example

File: src/View/StringTemplate.php [hunk 5]
```diff
@@ -323,4 +324,65 @@ protected function _formatAttribute($key, $value, $escape = true)
...
     }
+
+    /**
+     * Adds a class and returns a unique list either in array or space separated
+     *
+     * ### Options
+     *
+     * - `useIndex` if you are inputting an array with an 'element' called 'class' and want the same returning
+     *                  you should set this to 'class'
+     * - `returnArray` Setting this to `false` will return a space separated string (default is `true`)
```
Comment (dereuromark) [inline src/View/StringTemplate.php]: I am not sure, but I thought we used the other way around here in core: `'asString'` true/false. Defaulting to false.

**Comment by ADmad** [comment]
If the aim was to allowing passing options for labels around Radio and MultiCheckbox why did all widget classes have to be modified? The changes looks pretty invasive to me and are very likely to cause problems with app / plugin with custom/overridden widgets like friendofcake/bootstrap-ui.

Also the patch breaks existing tests.

**Comment by lilHermit** [comment]
@ADmad The other widgets were changed because `label` needs to be passed to the Widget and you would get `Label="1"` in the input's if not excluded.

I'm just fixing the two failed tests now.

**Comment by ADmad** [comment]
Well you shouldn't be passing unneeded options to other widgets. This change will definitely break custom widgets.

**Comment by ADmad** [comment]
Your previous patch #10100 has already caused problem with bootstrap ui plugin FriendsOfCake/bootstrap-ui#180 :)

**Comment by lilHermit** [comment]
@ADmad I'm happy to limit the extent `label` is passed to widgets (MultiCheckboxWidget & RadioWidget)

Re: #10100 it was a bug though as `nestedInput` was appearing in input rendering

**Comment by ADmad** [comment]
> I'm happy to limit the extent label is passed to widgets (MultiCheckboxWidget & RadioWidget)

Please do.

File: src/View/StringTemplate.php [hunk 6]
```diff
@@ -323,4 +324,65 @@ protected function _formatAttribute($key, $value, $escape = true)
...
+     *
+     * @param array|string $input The array or string to add the class to
+     * @param array|string $newClass the new class or classes to add
+     * @param array $options See above for options
+     * @return array|string
+     */
+    public function addClass($input, $newClass, $options = [])
```
Comment (ADmad) [inline src/View/StringTemplate.php]: For reference [OptionsAwareTrait](https://github.com/FriendsOfCake/bootstrap-ui/blob/master/src/View/Helper/OptionsAwareTrait.php) contains tried and tested code to add new class to existing list.

File: src/View/StringTemplate.php (hunk 2 above)
Comment (lilHermit) [inline src/View/StringTemplate.php]: Guards against `explode() expects parameter 2 to be string, object given`

Yes it's programmers error but thought it better to handle gracefully

File: src/View/StringTemplate.php (hunk 1 above)
Comment (lilHermit) [inline src/View/StringTemplate.php]: Yes that way it adds defaults and the options take precedence.

I've changed to

```php 
        $options += [
                'useIndex' => 'class',
                'asString' => false
            ];
```

File: tests/TestCase/View/Helper/FormHelperTest.php [hunk 7]
```diff
@@ -8496,4 +8498,266 @@ public function testNestedLabelInput()
...
+            'multiple' => 'checkbox',
+            'type' => 'select',
+            'options' => [
+                ['text' => 'First Checkbox', 'value' => 1],
+                ['text' => 'Second Checkbox', 'value' => 2]
+            ],
+            'optionsLabel' => ['class' => ['custom-class', 'another-class'], 'some-other-attr' => 'blah'],
+            'value' => ['1']
+
+        ]);
```
Comment (markstory) [inline tests/TestCase/View/Helper/FormHelperTest.php]: This is a really long test method. Could it be broken up into smaller tests that are co-located with the method under test?

File: src/View/Helper/FormHelper.php [hunk 8]
```diff
@@ -1008,7 +1008,8 @@ public function input($fieldName, array $options = [])
             'required' => null,
             'options' => null,
             'templates' => [],
-            'templateVars' => []
+            'templateVars' => [],
+            'optionsLabel' => true
```
Comment (lilHermit) [inline src/View/Helper/FormHelper.php]: Because the labels show as default, you can now pass `false` to skip the labels wrapping radio/checkboxes

File: src/View/Widget/MultiCheckboxWidget.php [hunk 9]
```diff
@@ -152,6 +153,9 @@ protected function _renderInputs($data, $context)
             if (!isset($checkbox['templateVars'])) {
                 $checkbox['templateVars'] = $data['templateVars'];
             }
+            if (!isset($checkbox['label'])) {
```
Comment (lilHermit) [inline src/View/Widget/MultiCheckboxWidget.php]: I was sticking with the same sanity check as `$checkbox['templateVars']` it stops the setting of $checkbox each loop of the foreach. Personally think its a bit pointless, I don't think its saving many cycles ;-)

File: tests/TestCase/View/Helper/FormHelperTest.php (hunk 7 above)
Comment (lilHermit) [inline tests/TestCase/View/Helper/FormHelperTest.php]: Yes I thought that too, I'll do so

File: src/View/Widget/MultiCheckboxWidget.php [hunk 10]
```diff
@@ -188,21 +192,31 @@ protected function _renderInput($checkbox, $context)
...
+                'text' => $checkbox['text'],
+                'templateVars' => $checkbox['templateVars'],
+                'input' => $input
+            ];
+
+            if (is_array($checkbox['label'])) {
+                $labelAttrs += $checkbox['label'];
+            }
+
+            if (isset($checkbox['checked']) && $checkbox['checked']) {
```
Comment (lilHermit) [inline src/View/Widget/MultiCheckboxWidget.php]: Agreed, that's now done

File: src/View/Helper/FormHelper.php (hunk 8 above)
Comment (ADmad) [inline src/View/Helper/FormHelper.php]: `labelOptions` would be a better name IMO.

Comment (lilHermit) [inline src/View/Helper/FormHelper.php]: ;-) I wanted to differentiate from the input label, and relates more to the options.

File: tests/TestCase/View/Helper/FormHelperTest.php [hunk 11]
```diff
@@ -8496,4 +8496,266 @@ public function testNestedLabelInput()
...
+            ['input' => ['type' => 'radio', 'name' => 'test', 'value' => '1', 'id' => 'test-1']],
+            '/div',
+        ];
+        $this->assertHtml($expected, $result);
+
+        // Tests String class (Radio)
+        $result = $this->Form->input('test', [
+            'type' => 'radio',
+            'options' => ['A', 'B'],
+            'optionsLabel' => ['class' => 'custom-class']
```
Comment (inoas) [inline tests/TestCase/View/Helper/FormHelperTest.php]: should be `labelOptions`

File: tests/TestCase/View/Helper/FormHelperTest.php [hunk 12]
```diff
@@ -8496,4 +8496,266 @@ public function testNestedLabelInput()
...
+            '/label',
+            '/div',
+        ];
+        $this->assertHtml($expected, $result);
+
+        // Tests class as array and adding selected as well as custom attributes (Radio)
+        $result = $this->Form->radio('test', ['A', 'B'], [
+            'label' => [
+                'class' => ['custom-class', 'another-class'],
+                'some-other-attr' => 'blah'
```
Comment (inoas) [inline tests/TestCase/View/Helper/FormHelperTest.php]: `data-*` attributes are more understandable.

File: src/View/Helper/FormHelper.php (hunk 8 above)
Comment (inoas) [inline src/View/Helper/FormHelper.php]: `optionsLabel` is even ambigious in a way. It could take an array of key/values where the key is the option value and the value is the option label (for radio/checkboxes) or <option>-tag conntent (for select/multi select).

**Comment by inoas** [comment]
> @ADmad I'm happy to limit the extent label is passed to widgets (MultiCheckboxWidget & RadioWidget)

`labelOptions` should imho apply to all checkbox (single, multiple) and radio controls to be consistent.

File: src/View/StringTemplate.php (hunk 6 above)
Comment (inoas) [inline src/View/StringTemplate.php]: Maybe that should be part of cake core then?

File: src/View/Widget/MultiCheckboxWidget.php (hunk 9 above)
Comment (markstory) [inline src/View/Widget/MultiCheckboxWidget.php]: Ignore me, I'm a muppet. I was misreading the diff and thought this chunk was connected to the chunk above that defaults `label` to `true`.

Comment (lilHermit) [inline src/View/Widget/MultiCheckboxWidget.php]: After further discussion I see it's there in case `$val` is an array

File: src/View/Helper/FormHelper.php (hunk 8 above)
Comment (lilHermit) [inline src/View/Helper/FormHelper.php]: `nestingLabel` / `nestedLabel` ?!

Comment (inoas) [inline src/View/Helper/FormHelper.php]: you are setting options so it should be `*options` ... and you are setting options for labels... so it should be labelsOptions or labelOptions... imho.

Comment (lilHermit) [inline src/View/Helper/FormHelper.php]: `labelOptions` then? 

(I'll make the change tomorrow and push all the other suggestions at the same time)

Comment (markstory) [inline src/View/Helper/FormHelper.php]: What about `labelAttributes` ?

Comment (lilHermit) [inline src/View/Helper/FormHelper.php]: The problem I have with `labelAttributes` is people might think it relates to the group label rather the one wrapping checkbox/radio. Happy to go with whatever though.

File: src/View/Helper/FormHelper.php [hunk 13]
```diff
@@ -1049,7 +1053,7 @@ public function input($fieldName, array $options = [])
             $options['hiddenField'] = '_split';
         }
 
-        $input = $this->_getInput($fieldName, $options);
+        $input = $this->_getInput($fieldName, $options + ['label' => $optionsLabel]);
```
Comment (ADmad) [inline src/View/Helper/FormHelper.php]: Why not keep the key `optionsLabel` / `labelOptions` here too?  Changing it to `label` is not intuitive.

File: src/View/Helper/FormHelper.php [hunk 14]
```diff
@@ -1850,7 +1856,7 @@ public function submit($caption = null, array $options = [])
 
         $input = $this->formatTemplate('inputSubmit', [
             'type' => $type,
-            'attrs' => $this->templater()->formatAttributes($options),
+            'attrs' => $this->templater()->formatAttributes($options, ['label']),
```
Comment (ADmad) [inline src/View/Helper/FormHelper.php]: Why is this change necessary?  Submit buttons have nothing to do with labels so that should never be passed to `submit()` options.

File: src/View/Helper/FormHelper.php (hunk 13 above)
Comment (lilHermit) [inline src/View/Helper/FormHelper.php]: I can change it

Comment (lilHermit) [inline src/View/Helper/FormHelper.php]: Once we have a decision on the name

File: src/View/Helper/FormHelper.php (hunk 14 above)
Comment (lilHermit) [inline src/View/Helper/FormHelper.php]: Because it was failing a test `testFormInputSubmit` due to the default method call in _getInput. I forgot to remove it, apologies!

File: src/View/Helper/FormHelper.php [hunk 15]
```diff
@@ -1930,6 +1936,8 @@ public function select($fieldName, $options = [], array $attributes = [])
             return $this->multiCheckbox($fieldName, $options, $attributes);
         }
 
+        unset($attributes['label']);
```
Comment (ADmad) [inline src/View/Helper/FormHelper.php]: This too is probably not necessary.

Comment (lilHermit) [inline src/View/Helper/FormHelper.php]: That is required to stop `label` getting to `SelectWidget` but still to `MultiCheckWidget`

Comment (lilHermit) [inline src/View/Helper/FormHelper.php]: 5 test failures without it as `label="1"` gets into `select`

Comment (markstory) [inline src/View/Helper/FormHelper.php]: So then it is necessary :smile:

File: src/View/Helper/FormHelper.php (hunk 8 above)
Comment (markstory) [inline src/View/Helper/FormHelper.php]: I'm good with `labelOptions` if you think 'attributes' will be confusing.

File: src/View/Helper/FormHelper.php [hunk 16]
```diff
@@ -1008,7 +1008,8 @@ public function input($fieldName, array $options = [])
             'required' => null,
             'options' => null,
             'templates' => [],
-            'templateVars' => []
+            'templateVars' => [],
+            'labelOptions' => true
```
Comment (ADmad) [inline src/View/Helper/FormHelper.php]: This new option should be documented in the method docblock.

File: src/View/Widget/MultiCheckboxWidget.php [hunk 17]
```diff
@@ -112,7 +112,8 @@ public function render(array $data, ContextInterface $context)
             'disabled' => null,
             'val' => null,
             'idPrefix' => null,
-            'templateVars' => []
+            'templateVars' => [],
+            'label' => true
```
Comment (ADmad) [inline src/View/Widget/MultiCheckboxWidget.php]: Why is this key `label` and not `labelOptions` to maintain consistency? :)

Comment (lilHermit) [inline src/View/Widget/MultiCheckboxWidget.php]: Because I deemed it not important within the widget as within that scope it's label

Comment (ADmad) [inline src/View/Widget/MultiCheckboxWidget.php]: The value of `label` would be the same as value of `labelOptions` passed to FormHelper::input() so better to keep the name same.

Comment (lilHermit) [inline src/View/Widget/MultiCheckboxWidget.php]: It has to remain as `label` in `RadioWidget` as it can be called direct from `FormHelper::radio`. Also you could call `FormHelper::multiCheckbox with `label` and it would work. I think it needs to remain as `label` in the widgets for these direct method functionality too.

Comment (lilHermit) [inline src/View/Widget/MultiCheckboxWidget.php]:
```php
$this->Form->multiCheckbox(
    'checkbox1',
    [
        ['text' => 'First Checkbox', 'value' => 1],
        ['text' => 'Second Checkbox', 'value' => 2]
    ],
    [
        'multiple' => 'checkbox',
        'label' => ['class' => ['custom-class', 'another-class'], 'data-name' => 'bob'],
        'value' => ['1']
    ]
);
```
Maybe we should just `label` to the docblock of `multiCheckbox`

Comment (ADmad) [inline src/View/Widget/MultiCheckboxWidget.php]: Since `label` was already used in `FormHelper::radio()` I guess it's okay.

Comment (lilHermit) [inline src/View/Widget/MultiCheckboxWidget.php]: I'll add `label` to docblock for multiCheckbox

File: src/View/StringTemplate.php [hunk 18]
```diff
@@ -323,4 +324,68 @@ protected function _formatAttribute($key, $value, $escape = true)
...
     }
+
+    /**
+     * Adds a class and returns a unique list either in array or space separated
+     *
+     * ### Options
+     *
+     * - `useIndex` if you are inputting an array with an 'element' other than 'class'.
+     *              Also setting to 'false' will manipulate the whole array (default is 'class')
+     * - `asString` Setting this to `true` will return a space separated string (default is `false`)
```
Comment (markstory) [inline src/View/StringTemplate.php]: Do we really need to offer array & string based return values? Can we get away with a single return type?

Comment (lilHermit) [inline src/View/StringTemplate.php]: This method comes from a Utility class I have (slightly modified). Ideally this method would live in HtmlHelper or a Html Utility class like `Cake\Utility\Text` so it can be used by end-user. Preferable a static Utility class then its easier to use in Core. With this in mind I think `asString` would be useful. However I'm happy to remove the option in this context.

Comment (dereuromark) [inline src/View/StringTemplate.php]: I also think we should keep the API here simple, and rather manually `implode(' ', $array)` afterwards, easy enough, right?

Comment (lilHermit) [inline src/View/StringTemplate.php]: alo `implode` is quite tricky ;-)

Comment (lilHermit) [inline src/View/StringTemplate.php]: Ok that's pushed. I also changed `$options` to `$useIndex` string param as we only have one option ;-)

File: src/View/StringTemplate.php [hunk 19]
```diff
@@ -323,4 +324,58 @@ protected function _formatAttribute($key, $value, $escape = true)
...
+            }
+        }
+
+        if (is_string($newClass)) {
+            $newClass = explode(' ', $newClass);
+        }
+
+        $class = array_unique(array_merge($class, $newClass));
+
+        if (is_string($useIndex)) {
```
Comment (markstory) [inline src/View/StringTemplate.php]: When would useIndex not be a string?

Comment (lilHermit) [inline src/View/StringTemplate.php]: If you are passing in an array without associative key as it supports `false` For example
```php
class MyClass
{
    private $defaultClasses = ['class1', 'class2'];
    private $defaultClassesString = 'class1 class2';

    private function myMethod()
    {
        $classes = $this->template->addClass(
            $this->defaultClasses,
            'new_class',
            false
        );
        // or
        $classes = $this->template->addClass(
            $this->defaultClassesString,
            'new_class',
            false
        );

    }
}
```

Comment (markstory) [inline src/View/StringTemplate.php]: Do we currently have a use case for that?

Comment (lilHermit) [inline src/View/StringTemplate.php]: Currently not but surely it might be useful!?! I don't see a reason not too for one line of code!!

Comment (markstory) [inline src/View/StringTemplate.php]: If we don't need it *now*, I prefer to not have 'extra' code. It adds complexity, requires more testing and more documentation.

File: src/View/Helper/FormHelper.php [hunk 20]
```diff
@@ -1463,7 +1472,9 @@ public function checkbox($fieldName, array $options = [])
      * ### Attributes:
      *
      * - `value` - Indicates the value when this radio button is checked.
-     * - `label` - boolean to indicate whether or not labels for widgets should be displayed.
+     * - `label` - Either `false` to disable label around the widget or an array of attributes for
+     *    the label tag. `selected` will be added to any classes e.g. `class => 'myclass'` where widget
```
Comment (markstory) [inline src/View/Helper/FormHelper.php]: Missing `'` around 'class' here.

File: src/View/Helper/FormHelper.php [hunk 21]
```diff
@@ -1970,6 +1983,9 @@ public function select($fieldName, $options = [], array $attributes = [])
      *   You can also set disabled to a list of values you want to disable when creating checkboxes.
      * - `hiddenField` - Set to false to remove the hidden field that ensures a value
      *   is always submitted.
+     * - `label` - Either `false` to disable label around the widget or an array of attributes for
+     *   the label tag. `selected` will be added to any classes e.g. `class => 'myclass'` where
```
Comment (markstory) [inline src/View/Helper/FormHelper.php]: Missing `'` around 'class' here.

File: src/View/Widget/RadioWidget.php [hunk 22]
```diff
@@ -173,6 +173,13 @@ protected function _renderInput($val, $text, $data, $context)
         if (isset($data['val']) && (string)$data['val'] === (string)$radio['value']) {
             $radio['checked'] = true;
         }
+
+        if (!is_bool($data['label'])) {
+            if (isset($radio['checked']) && $radio['checked']) {
```
Comment (markstory) [inline src/View/Widget/RadioWidget.php]: These conditions could be combined as there are no elses.

File: tests/TestCase/View/Helper/FormHelperTest.php [hunk 23]
```diff
@@ -8496,4 +8496,278 @@ public function testNestedLabelInput()
...
+            'label' => 'My checkboxes',
+            'multiple' => 'checkbox',
+            'type' => 'select',
+            'options' => [
+                ['text' => 'First Checkbox', 'value' => 1],
+                ['text' => 'Second Checkbox', 'value' => 2]
+            ],
+            'labelOptions' => ['class' => ['custom-class', 'another-class'], 'data-name' => 'bob'],
+            'value' => ['1']
+
```
Comment (markstory) [inline tests/TestCase/View/Helper/FormHelperTest.php]: Spare newline.

**Review by markstory — COMMENTED** [review]
If you could rebase this, I could get it merged in.

**Comment by markstory** [comment]
Merged into master in b138621cb2d99db0e4ebe7cff666f524473a04ba
